api/include/opentelemetry/trace/span.h

```diff
@@ -1,38 +1,112 @@
 #pragma once
 
+#include <cstdint>
+
 #include "opentelemetry/core/timestamp.h"
+#include "opentelemetry/nostd/string_view.h"
 
 namespace opentelemetry
 {
 namespace trace
 {
+enum class SpanKind
+{
+  kInternal,
+  kServer,
+  kClient,
+  kProducer,
+  kConsumer,
+};
 /**
- * StartSpanOptions provides options to set properties of span at the time of starting a new span.
+ * StartSpanOptions provides options to set properties of a Span at the time of its creation
  */
 struct StartSpanOptions
 {
-  // Sets the start time of a span.
+  // Optionally sets the start time of a Span.
   //
-  // If the start time of a span is set, timestamps from both the system clock and steady clock
-  // should be provided.
+  // If the start time of a Span is set, timestamps from both the system clock and steady clock
+  // must be provided.
   //
-  // Timestamps from the steady clock can be used to most accurately measure a span's
-  // duration, while timestamps from the system clock can be used to most accurately place a span's
-  // time point relative to other spans collected across a distributed system.
+  // Timestamps from the steady clock can be used to most accurately measure a Span's
+  // duration, while timestamps from the system clock can be used to most accurately place a Span's
+  // time point relative to other Spans collected across a distributed system.
   core::SystemTimestamp start_system_time;
   core::SteadyTimestamp start_steady_time;
+
+  // TODO:
+  // Span(Context?) parent;
+  // SpanContext remote_parent;
+  // Links
+  // Attributes
+  SpanKind kind = SpanKind::kInternal;
 };
 
 class Tracer;
 
 /**
- * A span represents a single operation within a trace.
+ * A Span represents a single operation within a Trace.
+ *
+ * Spans *must* be ended by calling End(). Merely deleting a Span does not end
+ * it.
```

Comment:
Why not have the destructor End?

Reply:
A Span can be created without having to start.
In Python we had similar discussion and decided to give flexibility https://github.com/open-telemetry/opentelemetry-python/blob/aecc3477b0431821015a0455492764baf90a0c36/opentelemetry-api/src/opentelemetry/trace/__init__.py#L591.

Reply:
In Java, End() has to be called explicitly. Do we want to make C++ different, or do we want to add a e.g. ScopedSpan wrapper that calls End?

Reply:
Requiring an explicit call to `End()` is an alternative to ownership semantics or refcounting. What about the case where multiple threads are working on the same Span, via the per-thread Context?

We can still add a ScopedSpan wrapper that calls End in the destructor.

Reply:
> What about the case where multiple threads are working on the same Span, via the per-thread Context?

Not sure I understand this case.

The way I did this is to use an atomic that ensures a span is only ended once. Like [this](https://github.com/lightstep/lightstep-tracer-cpp/blob/master/src/tracer/span.cpp#L235) so that if it's already ended it does nothing (and you can still call explicitly).

I don't know that Java makes a good comparison since it doesn't have destructors.

Reply:
>only ended once

Nice.

Separately from that, what's going to be the recommended way to have multiple threads or callbacks share a Span?

Reply:
I'd be for making the span api thread-safe (using either a mutex or spin lock)

Reply:
Per discussion in this week's meeting, destructor ends the span.